### ISSUE TITLE
Thoughts after the web-app-fundamentals premiere

### ISSUE THREAD

**Issue by onemen** [epicweb-dev/kcdshop#22]
## General

More information about the workshop app is needed, you can add a another read-me
page with more information about the workshop app, including screenshot,
explanation about navigation, tab panels, work flow etc...

Expand the explanation regarding the `exercises`, the folders structure, the
division of each exercise into lessons. Explain that each lesson contain a
problem folder which is the `working directory` for the student to work in and a
`teacher's solution` folder solution folder.

Perhaps you should add a note at the bottom of each lesson about closing the
files from previous lessons before starting to work on the current lesson (it
can be confusing especially when working on the same files in consecutive
lessons)

## Diff

At the beginning of each lesson the `DIFF` tab show the different of each file
touched in the lesson between the problem and the solution. however as the
student work progress in the lesson the diff tab losing its starting data, and
only show the remaining changes.

I think you can add 2nd diff tab with fixed diff for that lesson, one tab can
read as `lesson diff` and the other `current diff`. you can generate the fixed
diff from a GitHub action and the save the diff file/s in the solution folder.

Another thing that i find confusing is the possibility to view backward diff
between the solution to the problem. You can get such diff when you are on
solution view `http://localhost:5639/08/01/solution?preview=diff` or when
student select APP1 and APP2 in reverse order. I think that this can be
confusing and serve no real purpose. I suggest to always show the diff in
forward order.

## Build

The change in 807f349fb3a193ba06e3847a0aedb5592eb743a5 add build step to every
app in the setup stage. After this change the total time of a clean setup in
about `8 minutes`

| step                  | time (m:s) |
| --------------------- | ---------- |
| installing dependency | 1:17       |
| prisma & .env         | 0:35       |
| playwright            | 0:35       |
| build apps            | 5:30       |
| total time            | 7:57       |

Is it possible to generate this files (build, public/build) using GitHub action
and include the files in the repository?

I did not noticed any improvement in the start app in each lesson. 

## Bugs

- Apps start many time before the styles loaded. styles load after refresh.
  https://github.com/epicweb-dev/kcdshop/commit/fbd3fadd7cc8caee119ee9de84dcfc7a15f34df9
  did not help

- You mention during the workshop that the refresh button on the iFrame app
  should clear address bar. for example from `/settings/profile` to `/`. but
  this is not the normal behavior of a browser, the address bar does not reset
  after refresh. solution to this can be to

  - add `home` button
  - add button with the label 'RESET', that will restart/reset the app.

## UI improvement

- option to open all touched files in a lesson.

- you already mension about adding indication which app is running, it can be
  added on the navigation side bar and on the tabs `PROBLEM` `SOLUTION` of each
  lesson.

- option to close all apps, previous apps.

- option to automatically start both problem and solution apps of a lesson.

- option to automatically close previous apps when starting new lesson.

**Comment by kentcdodds** [comment]
Thanks for this!

I agree some instruction on how to use the workshop app would be helpful. I failed to add you to the calendar event which included a loom video demonstrating how to use the app. I also did give an introduction to the workshop app at the start before I realized I hadn't gotten you the zoom link yet.

> Perhaps you should add a note at the bottom of each lesson about closing the files from previous lessons before starting to work on the current lesson

I think #25 should address this

> generate the fixed diff

The "Files" menu at the bottom of the exercise is supposed to serve as the "Here are the files you need to work in" bit. The purpose of the diff is simply to help people if they get stuck. Most of the time people should not be using the `diff` so giving them a chance to just look at the diff more easily is not a priority and adding more tabs or options I think would be more confusing.

> possibility to view backward diff

This functionality is intended. When you're on the solution page, I think it makes sense that the solution would be the app number 1. Again, the diff feature isn't one I hope people to not use very often, so I'm not concerned about this as much.

> Is it possible to generate this files

The problem with generating files (especially build files) is that any change to those files will show up in the git diff which will annoy people working on the workshop. There could be another way to address this (like run the build script before running the dev script). I could also just remove running the build scripts. I didn't notice any useful difference there either. I made the change because someone complained about it, but I don't think it was a real issue.

> Apps start many time before the styles loaded.

I was hoping running the builds ahead of time would help with this. Can you confirm that it did not?

> You mention during the workshop that the refresh button on the iFrame app should clear address bar.

I don't remember saying this. I do remember that sometimes hitting the refresh button would show the home screen instead of the route that was in the URL bar which is definitely a bug. I'm not certain what caused it, but I would like to get that fixed. I don't think adding a `home` or `reset` button is necessary.

> option to open all touched files in a lesson.

GREAT idea 👍👍

> indication which app is running

Yup, this would be great 👍

**Comment by kentcdodds** [comment]
I think we've addressed (or are addressing) all of these points. Thanks!
